### PR TITLE
Show site origin in file select dialog

### PR DESCRIPTION
**Pull request by sangwoo108**
For security reason, we should show the site origin in the file select dialog

Resolves https://github.com/brave/brave-browser/issues/39682

### For open
image

### For save
image

## Submitter Checklist:

- [ ] I confirm that no [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) and no other type of reviews are needed, or that I have [requested](https://github.com/brave/reviews/issues/new/choose) them
- [ ] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [ ] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [ ] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [ ] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### For file open
* open inspector > console and type this (or you can use any sites that uploads files)
```
window.setTimeout(() => {
var input = document.createElement('input');
input.type = 'file';
input.click();
}, 2345);
```
* See if file chooser shows site origin in the title

### For file save
* Try downloading something, right click image or link and click `Save as ...`
* See if file chooser has site origin in the title

### CODE REVIEW

**Comment by bcaller** [comment]
I like it.

I will mention here that if the domain name is really really long we do truncate from the right.

This was discussed https://github.com/brave/brave-core/pull/24293#issuecomment-2185325843

e.g.
`http://xxxxxxxxxx.xxxxxxxxxx.xxxxxxxxxx.xxxxxxxxxx.xxxxxxxxxx.xxxxxxxxxx.xxxxxxxxxx.xxxxxxxxxx.xxxxxxxxxx.xxxxxxxxxx.xxxxxxxxxx.abcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijk.com/`

For file chooser it just shows the subdomain part.

Brave Browser Development 2024-07-12 22 33 19

Brave Browser Development 2024-07-12 22 33 38

But I don't think we have sufficient control of the dialog to fix this without a lot of extra work.

I have only tested on MacOS.

**Comment by bcaller** [comment]
It could help to see screenshots from other platforms for this

**Review by bcaller — APPROVED** [review]

**Comment by sangwoo108** [comment]
I found that on Windows, it doesn't work. Let me check a bit further. 

(I'm not able to check Linux at this moment, maybe we should get help from the QA team.)

**Comment by sangwoo108** [comment]
I found that upstream folks intentionally ignore title on Windows. Let me ask them.

File: browser/ui/brave_file_select_helper_unittest.cc [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+
+  // Whether the main frame is alerting.
+  bool is_main_frame;
+
+  // If `is_main_frame` is false, the URL of the alerting frame of the page.
+  const char* alerting_frame_url;
+
+  // The expected title for the alert.
+  const char* expected;
+} kCases[] = {
```
Comment (goodov) [inline browser/ui/brave_file_select_helper_unittest.cc:30]: pls make this a part of the test.

**Comment by sangwoo108** [comment]
They say 

> there is no reason why the title is ignored and that behavior originates from the initial git commit

I think it's safe to patch it.

File: browser/ui/brave_file_select_helper_unittest.cc (hunk 1 above)
Comment (sangwoo108) [reply browser/ui/brave_file_select_helper_unittest.cc:30]: Thanks, done

**Comment by sangwoo108** [comment]
### On Windows
image

image

File: app/brave_generated_resources.grd [hunk 2]
```diff
@@ -1299,6 +1299,32 @@ Or change later at <ph name="SETTINGS_EXTENIONS_LINK">$2<ex>brave://settings/ext
...
         </message>
       </if> <!-- enable_brave_player -->
+
+      <!-- File select -->
+      <message name="IDS_FILE_SELECT_OPEN_TITLE" desc="">
+        <ph name="SITE">$1<ex>http://foo.com</ex></ph> wants to open
+      </message>
+      <message name="IDS_FILE_SELECT_OPEN_TITLE_IFRAME" desc="">
+        An embedded page at <ph name="SITE">$1<ex>http://foo.com</ex></ph> wants to open
+      </message>
```
Comment (mkarolin) [inline app/brave_generated_resources.grd:1309]: I am not certain about `at` here and the other similar strings. I want to say `on` would sound more expected.

Comment (sangwoo108) [reply app/brave_generated_resources.grd]: Thanks, done!

**Review by mkarolin — APPROVED** [review]
`strings`++

**Comment by sangwoo108** [comment]
@goodov Could you take a look at this? Thanks!

File: browser/ui/brave_browser.cc [hunk 3]
```diff
@@ -128,6 +131,33 @@ void BraveBrowser::TabStripEmpty() {
...
+
+    // Note that save mode param is for PPAPI. 'Save As...' or downloading
+    // something doesn't reach here. They show 'select file dialog' from
+    // DownloadFilePicker::DownloadFilePicker directly.
+    // https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/public/mojom/choosers/file_chooser.mojom;l=27;drc=047c7dc4ee1ce908d7fea38ca063fa2f80f92c77
+
+    new_params->title = brave::GetFileSelectTitle(
+        content::WebContents::FromRenderFrameHost(render_frame_host),
+        render_frame_host->GetLastCommittedOrigin(),
+        params.mode == blink::mojom::FileChooserParams::Mode::kSave
```
Comment (goodov) [inline browser/ui/brave_browser.cc:153]: I think this need to be refactored into `bool is_save` or `enum class FileSelectTitleType { kOpen, kSave };`. This way you don't need to export ID maps, you can even store them as constexpr maps internally in `.cc`.

wdyt?

File: chromium_src/ui/shell_dialogs/execute_select_file_win.cc [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+
+// Can be removed once
+// https://chromium-review.googlesource.com/c/chromium/src/+/5711152 arrives in
+// Brave.
+#define BRAVE_EXECUTE_SELECT_MULTIPLE_FILE                               \
+  return RunOpenFileDialog(owner, title, std::u16string(), default_path, \
+                           filter, dialog_options, filter_index, paths);
+
+// Can be removed once
+// https://chromium-review.googlesource.com/c/chromium/src/+/5711152 arrives in
```
Comment (goodov) [inline chromium_src/ui/shell_dialogs/execute_select_file_win.cc:21]: nice job for creating and merging this into upstream!

File: patches/ui-shell_dialogs-execute_select_file_win.cc.patch [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
++  BRAVE_EXECUTE_SAVE_FILE
+   return RunSaveFileDialog(owner, std::u16string(), default_path, filter,
+                            dialog_options, def_ext, filter_index, path);
+ }
+@@ -421,7 +425,7 @@ void ExecuteSelectFile(
+       break;
+     case SelectFileDialog::SELECT_SAVEAS_FILE: {
+       base::FilePath path;
+-      if (ExecuteSaveFile(owner, default_path, filter, default_extension,
++      if (ExecuteSaveFile(owner, title, default_path, filter, default_extension,
```
Comment (goodov) [inline patches/ui-shell_dialogs-execute_select_file_win.cc.patch:41]: given that these patches will be eliminated when new Chromium will arrive, this is lgtm.

File: browser/ui/brave_file_select_helper_unittest.cc [hunk 6]
```diff
@@ -0,0 +1,213 @@
...
+    url::Origin alerting_frame_origin =
+        test_case.is_main_frame
+            ? main_frame_origin
+            : url::Origin::Create(GURL(test_case.alerting_frame_url));
+    EXPECT_EQ(
+        base::UTF8ToUTF16(test_case.expected),
+        brave::GetSiteFrameTitleForFileSelect(
+            brave::GetSiteFrameTitleType(main_frame_origin,
+                                         alerting_frame_origin),
+            alerting_frame_origin, brave::GetFileSelectResourceIDsForOpen()));
```
Comment (goodov) [inline browser/ui/brave_file_select_helper_unittest.cc:211]: is this missing `ForSave` tests?

also the locale should be enforced in the test suite when you're testing localizable strings. see `brave_l10n::test::ScopedDefaultLocale`.

File: browser/ui/brave_file_select_helper.h [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+}  // namespace url
+
+namespace brave {
+
+enum class SiteFrameTitleType {
+  kStandardSameOrigin,
+  kStandardDifferentOrigin,
+  kNonStandardSameOrigin,
+  kNonStandardDifferentOrigin
+};
```
[the remark this answers is not in the document]
Comment (sangwoo108) [reply browser/ui/brave_file_select_helper.h:28]: Good call. Basically this enum values represents what's done in `AppModalDialogManager::GetSiteFrameTitle`

File: browser/ui/brave_browser.cc (hunk 3 above)
Comment (sangwoo108) [reply browser/ui/brave_browser.cc:153]: I think having `enum class FileSelectTittleType` could be a good idea, but there's one more reason we have resource id maps parameter. As you suggested before, we need to compare `brave::GetFileSelectTitle` and `AppModalDialogManager::GetSiteFrameTitle`. So we should be able to specify resource ids to use.

File: chromium_src/ui/shell_dialogs/execute_select_file_win.cc (hunk 4 above)
Comment (sangwoo108) [reply chromium_src/ui/shell_dialogs/execute_select_file_win.cc:21]: Thanks! ☺️

File: browser/ui/brave_file_select_helper_unittest.cc (hunk 6 above)
Comment (sangwoo108) [reply browser/ui/brave_file_select_helper_unittest.cc:211]: Nice! Such a convenient function

File: browser/ui/brave_browser.cc (hunk 3 above)
Comment (goodov) [reply browser/ui/brave_browser.cc:153]: how about adding something like `FileSelectTittleType::kChromiumDefault` for tests?

File: app/brave_generated_resources.grd [hunk 8]
```diff
@@ -1299,6 +1299,32 @@ Or change later at <ph name="SETTINGS_EXTENIONS_LINK">$2<ex>brave://settings/ext
...
+      <message name="IDS_FILE_SELECT_SAVE_TITLE" desc="">
+        <ph name="SITE">$1<ex>http://foo.com</ex></ph> wants to save
+      </message>
+      <message name="IDS_FILE_SELECT_SAVE_TITLE_IFRAME" desc="">
+        An embedded page on <ph name="SITE">$1<ex>http://foo.com</ex></ph> wants to save
+      </message>
+      <message name="IDS_FILE_SELECT_SAVE_TITLE_NONSTANDARD_URL" desc="">
+        This page wants to save
+      </message>
+      <message name="IDS_FILE_SELECT_SAVE_TITLE_NONSTANDARD_URL_IFRAME" desc="">
```
Comment (goodov) [inline app/brave_generated_resources.grd:1325]: nit: these string IDs may conflict with chromium at some point. I'd use `IDS_BRAVE_...` here.

Comment (sangwoo108) [reply app/brave_generated_resources.grd]: Good point. Done.

File: browser/ui/brave_browser.cc (hunk 3 above)
Comment (sangwoo108) [reply browser/ui/brave_browser.cc:153]: That's a good idea! Done.

**Review by goodov — APPROVED** [review]

**Comment by GeetaSarvadnya** [comment]
Verification INPROGRESS on

```
Brave | 1.70.94 Chromium: 128.0.6613.114 (Official Build) beta (64-bit)
-- | --
Revision | 7ee7ec28264896bd8a9aff39cdb9432da38be028
OS | Windows 10 Version 22H2 (Build 19045.4780)
```

**Comment by AndyAnds265** [comment]
hey all, is there a UI screenshot associated with these strings? I'm trying to add it for context for the linguists!

**Comment by bsclifton** [comment]
@AndyAnds265 here are the two most popular ones

1. `<message name="IDS_BRAVE_FILE_SELECT_OPEN_TITLE" desc="">
    <ph name="SITE">$1<ex>http://foo.com</ex></ph> wants to open
</message>`

    image

2. `<message name="IDS_BRAVE_FILE_SELECT_SAVE_TITLE" desc="">
    <ph name="SITE">$1<ex>http://foo.com</ex></ph> wants to save
</message>`

    image


The others will look similar, but are just triggered differently. Ex: save triggrered from inside an iframe, etc.

**Comment by diracdeltas** [comment]
the strings were changed slightly in https://github.com/brave/brave-core/pull/27040/files for iframes btw

**Pull request merged**